### PR TITLE
ARROW-2400: [C++] Fix Status destructor performance

### PR DESCRIPTION
**Pull request by pitrou**
On certain compilers, splitting off the slow path improves performance significantly.

### CODE REVIEW

File: cpp/src/arrow/status.h [hunk 1]
```diff
@@ -94,8 +94,14 @@ class ARROW_MUST_USE_RESULT ARROW_EXPORT Status;
 class ARROW_EXPORT Status {
  public:
   // Create a success status.
-  Status() : state_(NULL) {}
-  ~Status() { delete state_; }
+  Status() noexcept : state_(NULL) {}
+  ~Status() noexcept {
+    // ARROW-2400: On certain compilers, splitting off the slow path improves
+    // performance significantly.
+    if (ARROW_PREDICT_FALSE(state_ != NULL)) {
```
Comment (cpcloud) [inline cpp/src/arrow/status.h]: Can we write this without the double negative? Does the branch prediction provide a material speedup? Or is it mostly from avoiding the destruction code + `noexcept`?

Comment (pitrou) [reply cpp/src/arrow/status.h]: Branch prediction is done by the CPU. Here we instruct the compiler that the `state_ != NULL` path is a cold path.

I'll remove the double negative.

Comment (pitrou) [reply cpp/src/arrow/status.h]: To answer your question, the speedup seems to stem from separating the fast path from the cold path.

Comment (pitrou) [reply cpp/src/arrow/status.h]: Hmm, removing the double negative makes the condition less readable IMHO (`!ARROW_PREDICT_TRUE(state_== NULL)`).

Comment (cpcloud) [reply cpp/src/arrow/status.h]: Cool, leaving it as is fine by me.

**Review by cpcloud — APPROVED** [review]
+1, LGTM

**Review by wesm — APPROVED** [review]
+1, thanks for looking at this @pitrou!

**Pull request closed without merging**